# HDB Resale Price Prediction

## Data Blending

### Import Libraries

In [1]:
# General
import math
import re
import numpy as np
import pandas as pd
from datetime import datetime
from tabulate import tabulate
from math import radians
from sklearn.metrics.pairwise import haversine_distances

# Warnings
import warnings
warnings.filterwarnings('ignore')

### Import HDB Dataset

In [2]:
# Importing Data
data_hdb_raw = pd.read_csv('../dataset/hdb_latest_raw.csv')
print(data_hdb_raw.shape)
data_hdb_raw.head()

(881924, 18)


Unnamed: 0        town flat_type flat_model  floor_area_sqm  \
0           0  ANG MO KIO    3 ROOM   Improved            69.0   
1           1  ANG MO KIO    3 ROOM   Improved            60.0   
2           2  ANG MO KIO    2 ROOM   Improved            45.0   
3           3  ANG MO KIO    2 ROOM   Improved            45.0   
4           4  ANG MO KIO    2 ROOM   Improved            45.0   

        street_name  resale_price    month  lease_commence_date storey_range  \
0  ANG MO KIO AVE 4      147000.0  2000-01                 1986     07 TO 09   
1  ANG MO KIO AVE 4      130000.0  2000-02                 1986     07 TO 09   
2  ANG MO KIO AVE 4      105000.0  2000-03                 1986     04 TO 06   
3  ANG MO KIO AVE 4      125000.0  2000-05                 1986     04 TO 06   
4  ANG MO KIO AVE 4      117000.0  2000-06                 1986     07 TO 09   

  block remaining_lease               address  \
0   170              85  170 ANG MO KIO AVE 4   
1   170              85  170 ANG MO KIO AVE 4   
2   170              85  170 ANG MO KIO AVE 4   
3   170              85  170 ANG MO KIO AVE 4   
4   170              85  170 ANG MO KIO AVE 4   

                                        full_address       lat        long  \
0  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
1  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
2  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
3  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
4  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   

  nearest_mrt  nearest_distance_to_mrt  
0   mayflower                    0.283  
1   mayflower                    0.283  
2   mayflower                    0.283  
3   mayflower                    0.283  
4   mayflower                    0.283

### Truncate Dataset

In [30]:
# Reduce to Last 15 Years i.e. 2009 onwards
data_hdb_last15 = data_hdb_raw
data_hdb_last15['month'] = pd.to_datetime(data_hdb_last15['month'])
data_hdb_last15 = data_hdb_last15[data_hdb_last15['month'].dt.year >= 2009]
data_hdb_last15.head()

Unnamed: 0        town flat_type flat_model  floor_area_sqm  \
77          77  ANG MO KIO    3 ROOM   Improved            61.0   
78          78  ANG MO KIO    3 ROOM   Improved            60.0   
79          79  ANG MO KIO    3 ROOM   Improved            69.0   
80          80  ANG MO KIO    2 ROOM   Improved            45.0   
81          81  ANG MO KIO    3 ROOM   Improved            61.0   

         street_name  resale_price      month  lease_commence_date  \
77  ANG MO KIO AVE 4      200000.0 2009-03-01                 1986   
78  ANG MO KIO AVE 4      200000.0 2009-04-01                 1986   
79  ANG MO KIO AVE 4      224000.0 2009-06-01                 1986   
80  ANG MO KIO AVE 4      170000.0 2009-07-01                 1986   
81  ANG MO KIO AVE 4      193000.0 2009-07-01                 1986   

   storey_range block remaining_lease               address  \
77     04 TO 06   170              76  170 ANG MO KIO AVE 4   
78     04 TO 06   170              76  170 ANG MO KIO AVE 4   
79     04 TO 06   170              76  170 ANG MO KIO AVE 4   
80     07 TO 09   170              76  170 ANG MO KIO AVE 4   
81     04 TO 06   170              76  170 ANG MO KIO AVE 4   

                                         full_address       lat        long  \
77  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
78  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
79  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
80  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   
81  170 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  1.374001  103.836432   

   nearest_mrt  nearest_distance_to_mrt  
77   mayflower                    0.283  
78   mayflower                    0.283  
79   mayflower                    0.283  
80   mayflower                    0.283  
81   mayflower                    0.283

### Data Wrangling

In [4]:
# Understanding Data I
# No columns with null entries
data_hdb_last15.info()

<class 'pandas.core.frame.DataFrame'>
Index: 325970 entries, 77 to 602220
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Unnamed: 0               325970 non-null  int64         
 1   town                     325970 non-null  object        
 2   flat_type                325970 non-null  object        
 3   flat_model               325970 non-null  object        
 4   floor_area_sqm           325970 non-null  float64       
 5   street_name              325970 non-null  object        
 6   resale_price             325970 non-null  float64       
 7   month                    325970 non-null  datetime64[ns]
 8   lease_commence_date      325970 non-null  int64         
 9   storey_range             325970 non-null  object        
 10  block                    325970 non-null  object        
 11  remaining_lease          325970 non-null  object        
 12  address             

In [31]:
# Drop Unnamed First (Index) Column
data_hdb_last15 = data_hdb_last15.drop(data_hdb_last15.columns[0], axis = 1)

In [32]:
# Merge flat model with different names that refer to the same type
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('MODEL A','Model A')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('IMPROVED','Improved')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('NEW GENERATION','New Generation')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('PREMIUM APARTMENT', 'Premium Apartment')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('SIMPLIFIED','Simplified')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('APARTMENT','Apartment')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('STANDARD','Standard')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('MAISONETTE','Maisonette')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('2-ROOM', '2-room')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('IMPROVED-MAISONETTE', 'Improved-Maisonette')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('MODEL A-MAISONETTE', 'Model A-Maisonette')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('MULTI GENERATION', 'Multi Generation')
data_hdb_last15['flat_model'] = data_hdb_last15['flat_model'].str.replace('TERRACE','Terrace')

In [33]:
# Finding out the different types of flat models for each flat type
grouped = data_hdb_last15.groupby('flat_type')['flat_model'].unique()
results = []

for flat_type, flat_models in grouped.items():
    results.append([flat_type, ', '.join(map(str, flat_models))])

print(tabulate(results, headers=["Flat Type", "Flat Models"], tablefmt="pretty"))

+------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------+
|    Flat Type     |                                                                     Flat Models                                                                     |
+------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------+
|      1 ROOM      |                                                                      Improved                                                                       |
|      2 ROOM      |                                            Improved, Standard, 2-room, Model A, Premium Apartment, DBSS                                             |
|      3 ROOM      |                              Improved, New Generation, Model A, Simplified, Standard, Terrace, Premium Apartment, DBSS      

In [34]:
# Combining 'flat_type' and 'flat_model' to give a better representation of the flat's category
data_hdb_last15['flat_category'] = data_hdb_last15['flat_type'] + " " + data_hdb_last15['flat_model']

In [35]:
# Change lease_commence_date to dt
data_hdb_last15['lease_commence_date'] = pd.to_datetime(data_hdb_last15['lease_commence_date'], format='%Y')

In [36]:
## USE THE CELLS HERE ##
# Source : https://hecksrealty.com/singapore-district-guide/
data_hdb_last15['postal_code'] = data_hdb_last15['full_address'].str[-6:]

def determine_region(row):
    postal_sector = str(row['postal_code'])[:2]

    if postal_sector in ['53', '54', '55', '82', '56', '57', '72', '73', '77', '78', '75', '76', '79', '80']:
        return 'North'
    elif postal_sector in ['14', '15', '16', '09', '10']:
        return 'South'
    elif postal_sector in ['34', '36', '36', '37',  '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '81', '51', '52']:
        return 'East'
    elif postal_sector in ['11', '12', '13', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71']:
        return 'West'
    elif postal_sector in ['20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33']:
        return 'Central' 
    elif postal_sector in ['01', '02', '03', '04', '05', '06', '07', '08', '17', '18', '19']:
        return 'City'
    else:
        return 'Others'

    
data_hdb_last15['Region'] = data_hdb_last15.apply(determine_region, axis=1)

print(data_hdb_last15)


              town flat_type flat_model  floor_area_sqm       street_name  \
77      ANG MO KIO    3 ROOM   Improved            61.0  ANG MO KIO AVE 4   
78      ANG MO KIO    3 ROOM   Improved            60.0  ANG MO KIO AVE 4   
79      ANG MO KIO    3 ROOM   Improved            69.0  ANG MO KIO AVE 4   
80      ANG MO KIO    2 ROOM   Improved            45.0  ANG MO KIO AVE 4   
81      ANG MO KIO    3 ROOM   Improved            61.0  ANG MO KIO AVE 4   
...            ...       ...        ...             ...               ...   
602216      YISHUN    2 ROOM    Model A            38.0      YISHUN AVE 4   
602217      YISHUN    4 ROOM    Model A            93.0      YISHUN AVE 4   
602218      YISHUN    3 ROOM    Model A            68.0      YISHUN AVE 4   
602219      YISHUN    5 ROOM   Improved           112.0      YISHUN AVE 4   
602220      YISHUN    5 ROOM   Improved           112.0      YISHUN AVE 4   

        resale_price      month lease_commence_date storey_range block  \
7

In [11]:
# Check Updated HDB Dataset
data_hdb_last15.info()

<class 'pandas.core.frame.DataFrame'>
Index: 325970 entries, 77 to 602220
Data columns (total 20 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   town                     325970 non-null  object        
 1   flat_type                325970 non-null  object        
 2   flat_model               325970 non-null  object        
 3   floor_area_sqm           325970 non-null  float64       
 4   street_name              325970 non-null  object        
 5   resale_price             325970 non-null  float64       
 6   month                    325970 non-null  datetime64[ns]
 7   lease_commence_date      325970 non-null  datetime64[ns]
 8   storey_range             325970 non-null  object        
 9   block                    325970 non-null  object        
 10  remaining_lease          325970 non-null  object        
 11  address                  325970 non-null  object        
 12  full_address        

In [12]:
# Unique street_names
print(data_hdb_last15['street_name'].nunique())
data_hdb_last15['street_name'].unique()

566


array(['ANG MO KIO AVE 4', 'ANG MO KIO AVE 1', 'ANG MO KIO AVE 10',
       'ANG MO KIO AVE 3', 'BEDOK NTH RD', 'ANG MO KIO AVE 5',
       'ANG MO KIO AVE 6', 'ANG MO KIO AVE 8', 'ANG MO KIO AVE 9',
       'ANG MO KIO AVE 2', 'ANG MO KIO ST 21', 'ANG MO KIO ST 31',
       'ANG MO KIO ST 11', 'BEDOK RESERVOIR RD', 'BEDOK NTH ST 3',
       'BEDOK STH AVE 1', 'BEDOK STH RD', 'CHAI CHEE AVE',
       'NEW UPP CHANGI RD', 'CHAI CHEE DR', 'CHAI CHEE RD',
       'BEDOK NTH AVE 1', 'BEDOK NTH AVE 2', 'BEDOK NTH AVE 3',
       'BEDOK NTH AVE 4', 'BEDOK NTH ST 1', 'BEDOK NTH ST 2',
       'BEDOK NTH ST 4', 'BEDOK STH AVE 2', 'BEDOK STH AVE 3',
       'CHAI CHEE ST', 'LENGKONG TIGA', 'JLN TENAGA', 'SIN MING RD',
       'SHUNFU RD', 'BISHAN ST 11', 'BISHAN ST 12', 'BISHAN ST 13',
       'SIN MING AVE', 'BISHAN ST 22', 'BISHAN ST 23', 'BISHAN ST 24',
       'BRIGHT HILL DR', 'BT BATOK ST 51', 'BT BATOK ST 52',
       'BT BATOK ST 11', 'BT BATOK ST 21', 'BT BATOK ST 34',
       'BT BATOK WEST AVE 4', 

The dataset has 566 unique street names. By grouping the HDB units by street name to calculate the average coordinate of hdb units in the street, we can significantly cut down the time required to calculate the number of POIs within a certain radius of that average coordinate.

In [13]:
# getting average coordinates of each street name
street_avg_coords = data_hdb_last15.groupby('street_name').agg({'long': 'mean', 'lat': 'mean'}).reset_index()
street_avg_coords.columns = ['street_name', 'avg_long', 'avg_lat']

#ave_coord_df = pd.merge(data_hdb_last15, street_avg_coords, on='street_name', how='left')
#ave_coord_df
street_avg_coords

street_name    avg_long   avg_lat
0      ADMIRALTY DR  103.815851  1.450018
1    ADMIRALTY LINK  103.816254  1.455540
2        AH HOOD RD  103.844667  1.327701
3    ALJUNIED AVE 2  103.886335  1.319203
4     ALJUNIED CRES  103.884427  1.320423
..              ...         ...       ...
561   YUNG KUANG RD  103.721629  1.331872
562     YUNG LOH RD  103.722128  1.328313
563    YUNG PING RD  103.722079  1.329695
564   YUNG SHENG RD  103.721681  1.333492
565         ZION RD  103.832608  1.290660

[566 rows x 3 columns]

### Import POI Dataset

In [14]:
# Importing Data
data_poi_raw = pd.read_csv('../dataset/points_of_interest.csv')
data_poi_raw

Unnamed: 0                     place_id                         name  \
0              0  ChIJ01fgzLUe2jERxlhvImcbZ7g                Quayside Isle   
1              1  ChIJ1S4qfY8Q2jERgb68gskzUbo            Sime Darby Centre   
2              2  ChIJ1ZAIkrwZ2jERxtZGC1JnrHM                         PoMo   
3              3  ChIJ1ZYJOiAZ2jER1mvQqHstQII               LR boulangerie   
4              4  ChIJ2Y1DYBI92jERlFUKKSznJrY                 Tampines Hub   
...          ...                          ...                          ...   
8667        8667  ChIJic0FbG4T2jERlg__g9KNSUU  Old Chang Kee (Head Office)   
8668        8668  ChIJ7cjSMX0T2jERGYqGQog7A2E    Old Chang Kee @ Sun Plaza   
8669        8669  ChIJxXwRE24T2jERDGyVjkkHTTs           Old Chang Kee Bldg   
8670        8670  ChIJmQOMh1AT2jERs_1tteD7eTg      Old Chang Kee Coldstore   
8671        8671  ChIJkYvJbm4T2jERIfS4wTtaBZg    Ten & Han Trading Pte Ltd   

           lat         lng  rating  user_ratings_total  price_level  \
0     1.247681  103.842072     4.3               568.0          NaN   
1     1.336644  103.783597     3.7               437.0          NaN   
2     1.300192  103.849220     3.8              1285.0          NaN   
3     1.293178  103.827194     4.3                12.0          NaN   
4     1.353108  103.940361     4.6               227.0          NaN   
...        ...         ...     ...                 ...          ...   
8667  1.450192  103.805305     4.3                25.0          1.0   
8668  1.448144  103.819983     3.4                10.0          1.0   
8669  1.449830  103.805229     4.0                 2.0          NaN   
8670  1.468055  103.812869     0.0                 0.0          NaN   
8671  1.450047  103.805526     3.0                 1.0          NaN   

                                      formatted_address  global_code  ...  \
0                        31 Ocean Way, Singapore 098375  6PH56RXR+3R  ...   
1                      896 Dunearn Rd, Singapore 589472  6PH58QPM+MC  ...   
2                        1 Selegie Rd, Singapore 188306  6PH58R2X+3M  ...   
3     491 River Valley Rd, #01-02 valley point shopp...  6PH57RVG+7V  ...   
4                     1 Tampines Walk, Singapore 528523  6PH59W3R+64  ...   
...                                                 ...          ...  ...   
8667                     2 Woodlands Terrace, Singapore  6PH5FR24+34  ...   
8668       30 Sembawang Dr, #B1-44 Sun Plaza, Singapore  6PH5CRX9+7X  ...   
8669                                          Singapore  6PH5CRX4+W3  ...   
8670                           20 Senoko Way, Singapore  6PH5FR97+64  ...   
8671                     2 Woodlands Terrace, Singapore  6PH5FR24+26  ...   

     train_station natural_feature subpremise  SUBZONE_NO          SUBZONE_N  \
0            False           False      False         1.0            SENTOSA   
1            False           False      False         2.0         SWISS CLUB   
2            False           False      False         8.0            SELEGIE   
3            False           False      False         2.0         CHATSWORTH   
4            False           False      False         3.0      TAMPINES WEST   
...            ...             ...        ...         ...                ...   
8667         False           False      False         9.0       SENOKO SOUTH   
8668         False           False      False         3.0  SEMBAWANG CENTRAL   
8669         False           False      False         9.0       SENOKO SOUTH   
8670         False           False      False         8.0       SENOKO NORTH   
8671         False           False      False         9.0       SENOKO SOUTH   

      SUBZONE_C        PLN_AREA_N  PLN_AREA_C        REGION_N  REGION_C  
0        SISZ01  SOUTHERN ISLANDS          SI  CENTRAL REGION        CR  
1        BTSZ02       BUKIT TIMAH          BT  CENTRAL REGION        CR  
2        RCSZ08            ROCHOR          RC  CENTRAL REGION        CR  
3        TNSZ02           TANGLIN 

In [15]:
# Understanding Data II
data_poi_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8672 entries, 0 to 8671
Columns: 125 entries, Unnamed: 0 to REGION_C
dtypes: bool(105), float64(6), int64(1), object(13)
memory usage: 2.2+ MB


In [16]:
# Understanding Data III
for i in range(len(data_poi_raw.columns)):
    print(data_poi_raw.columns[i])

Unnamed: 0
place_id
name
lat
lng
rating
user_ratings_total
price_level
formatted_address
global_code
compound_code
planning_area
brand
establishment
point_of_interest
store
food
health
restaurant
hospital
lodging
finance
cafe
convenience_store
clothing_store
atm
shopping_mall
grocery_or_supermarket
home_goods_store
school
bakery
beauty_salon
transit_station
place_of_worship
pharmacy
meal_takeaway
furniture_store
tourist_attraction
secondary_school
supermarket
doctor
shoe_store
dentist
jewelry_store
church
bank
primary_school
electronics_store
gym
spa
car_repair
pet_store
bus_station
university
park
general_contractor
subway_station
real_estate_agency
florist
hair_care
department_store
hardware_store
car_dealer
veterinary_care
travel_agency
bicycle_store
book_store
laundry
plumber
meal_delivery
lawyer
parking
mosque
physiotherapist
art_gallery
insurance_agency
bar
museum
storage
movie_theater
moving_company
liquor_store
gas_station
electrician
car_rental
locksmith
car_wash
post_office
e

In [19]:
# Retain POI types related to health, recreational areas and schools
columns_to_keep = ['name', 'rating', 'lat', 'lng', 'hospital', 'doctor', 'dentist', 'park', 'shopping_mall', 'amusement_park', 'zoo', 'library',
                   'school', 'primary_school', 'secondary_school', 'university']

data_poi_filtered = data_poi_raw[columns_to_keep]

# Drop rows where all selected POI types are False
data_poi_filtered = data_poi_filtered[data_poi_filtered.iloc[:, 3:].any(axis=1)]

# Create new columns for POI archetypes
data_poi_filtered['Healthcare'] = data_poi_filtered[['hospital', 'doctor', 'dentist']].any(axis=1)
data_poi_filtered['Recreational'] = data_poi_filtered[['park', 'shopping_mall', 'amusement_park', 'zoo', 'library']].any(axis=1)
data_poi_filtered['Education'] = data_poi_filtered[['school', 'primary_school', 'secondary_school', 'university']].any(axis=1)

data_poi_filtered = data_poi_filtered[['name', 'rating', 'lat', 'lng', 'Healthcare', 'Recreational', 'Education']]

data_poi_filtered

name  rating       lat         lng  Healthcare  \
0                   Quayside Isle     4.3  1.247681  103.842072       False   
1               Sime Darby Centre     3.7  1.336644  103.783597       False   
2                            PoMo     3.8  1.300192  103.849220       False   
3                  LR boulangerie     4.3  1.293178  103.827194       False   
4                    Tampines Hub     4.6  1.353108  103.940361       False   
...                           ...     ...       ...         ...         ...   
8667  Old Chang Kee (Head Office)     4.3  1.450192  103.805305       False   
8668    Old Chang Kee @ Sun Plaza     3.4  1.448144  103.819983       False   
8669           Old Chang Kee Bldg     4.0  1.449830  103.805229       False   
8670      Old Chang Kee Coldstore     0.0  1.468055  103.812869       False   
8671    Ten & Han Trading Pte Ltd     3.0  1.450047  103.805526       False   

      Recreational  Education  
0             True      False  
1             True      False  
2             True      False  
3            False      False  
4             True      False  
...            ...        ...  
8667         False      False  
8668         False      False  
8669         False      False  
8670         False      False  
8671         False      False  

[8672 rows x 7 columns]

### Merge Datasets

In [25]:
# Function to calculate Haversine distance between two points
def haversine_distance(lng1, lat1, lng2, lat2):
    lng1, lat1, lng2, lat2 = map(radians, [lng1, lat1, lng2, lat2])
    distances = haversine_distances([[lat1, lng1], [lat2, lng2]])
    return distances[1, 0] * 6371.0  # multiply by Earth radius to get km

# Function to count POIs and return average rating of a certain type within a specified radius from given long and lat
def count_pois_within_radius(hdb_lng, hdb_lat, poi_data, poi_type, radius):
    count = 0
    total_rating = 0;
    for index, poi in data_poi_filtered.iterrows():
        poi_lng = poi['lng']
        poi_lat = poi['lat']
        distance = haversine_distance(hdb_lng, hdb_lat, poi_lng, poi_lat)
        if distance <= radius and poi[poi_type] == True:
            count += 1
            total_rating += poi['rating']

    if(count == 0):
        average_rating = 0
    else:
        average_rating = total_rating / count
        
    return count, average_rating

In [23]:
# To test below function with a smaller subset of data
hdb_resale_data = data_hdb_last15.sample(n = 10)
hdb_resale_data

town flat_type         flat_model  floor_area_sqm  \
486248  JURONG WEST    4 ROOM           Model A2            85.0   
10225    ANG MO KIO    3 ROOM     New Generation            89.0   
176935     CLEMENTI    5 ROOM           Improved           118.0   
285342   ANG MO KIO    5 ROOM           Improved           117.0   
561239  JURONG WEST    5 ROOM  Premium Apartment           110.0   
399890    WOODLANDS    4 ROOM            Model A           100.0   
240618  BUKIT BATOK    4 ROOM         Simplified            84.0   
414569     TAMPINES    4 ROOM            Model A           104.0   
596264    TOA PAYOH    4 ROOM            Model A            93.0   
545453      PUNGGOL    5 ROOM           Improved           110.0   

               street_name  resale_price      month lease_commence_date  \
486248   JURONG WEST ST 64      331800.0 2009-03-01          2000-01-01   
10225    ANG MO KIO AVE 10      380000.0 2015-10-01          1979-01-01   
176935  CLEMENTI WEST ST 2      670000.0 2021-04-01          1981-01-01   
285342    ANG MO KIO AVE 4      420000.0 2009-11-01          1978-01-01   
561239   JURONG WEST ST 93      448000.0 2019-04-01          2008-01-01   
399890     WOODLANDS DR 40      292500.0 2009-12-01          1996-01-01   
240618      BT BATOK ST 32      304000.0 2019-10-01          1987-01-01   
414569      TAMPINES ST 71      441000.0 2011-08-01          1997-01-01   
596264   UPP ALJUNIED LANE      720000.0 2021-10-01          2017-01-01   
545453       PUNGGOL FIELD      445000.0 2018-07-01          2003-01-01   

       storey_range block     remaining_lease                 address  \
486248     07 TO 09  668B                  90  668B JURONG WEST ST 64   
10225      04 TO 06   416                  62   416 ANG MO KIO AVE 10   
176935     16 TO 18   724  58 years 09 months  724 CLEMENTI WEST ST 2   
285342     04 TO 06   111                  68    111 ANG MO KIO AVE 4   
561239     04 TO 06  988A  88 years 04 months  988A JURONG WEST ST 93   
399890     01 TO 03   702                  86     702 WOODLANDS DR 40   
240618     01 TO 03   310  66 years 04 months      310 BT BATOK ST 32   
414569     04 TO 06   725                  85      725 TAMPINES ST 71   
596264     10 TO 12     7  94 years 05 months     7 UPP ALJUNIED LANE   
545453     07 TO 09  173A  84 years 06 months      173A PUNGGOL FIELD   

                                             full_address       lat  \
486248        668B JURONG WEST STREET 64 SINGAPORE 642668  1.341928   
10225   416 ANG MO KIO AVENUE 10 TECK GHEE HEARTLANDS ...  1.364542   
176935        724 CLEMENTI WEST STREET 2 SINGAPORE 120724  1.303609   
285342  111 ANG MO KIO AVENUE 4 KEBUN BARU NEIGHBOURHO...  1.370816   
561239        988A JURONG WEST STREET 93 SINGAPORE 641988  1.336487   
399890  702 WOODLANDS DRIVE 40 HDB PUBLIC SHELTERS SIN...  1.441524   
240618         310 BUKIT BATOK STREET 32 SINGAPORE 650310  1.359641   
414569  725 TAMPINES STREET 71 PCF SPARKLETOTS PRESCHO...  1.356985   
596264  7 UPPER ALJUNIED LANE JOO SENG GREEN SINGAPORE...  1.334715   
545453                173A PUNGGOL FIELD SINGAPORE 821173  1.397013   

              long   nearest_mrt  nearest_distance_to_mrt  \
486248  103.703386      boon lay                    0.475   
10225   103.855104    ang mo kio                    0.860   
176935  103.763359      clementi                    1.296   
285342  103.836984     mayflower                    0.086   
561239  103.695615       pioneer                    0.226   
399890  103.796664     admiralty                    0.492   
240618  103.748829  bukit gombak                    0.349   
414569  103.935335      tampines                    1.165   
596264  103.878708       bartley                    0.881   
545453  103.909354      meridian                    0.046   

                   flat_category postal_code Region  
486248           4 ROOM Model A2      642668   West  
10225      3 ROOM New Generation      560416  North  
176935           5 R

In [26]:
# Iterate through each unqiue Street Name as a proxy for the lat & long of the hdb unit
# Calculate the average rating for each POI archetype
for index, hdb_unit in street_avg_coords.iterrows():
    hdb_lng = hdb_unit['avg_long']
    hdb_lat = hdb_unit['avg_lat']

    healthcare_within_1km_count = 0
    recreational_within_1km_count = 0
    education_within_1km_count = 0
    healthcare_within_2km_count = 0
    recreational_within_2km_count = 0
    education_within_2km_count = 0
    healthcare_within_1km_average_rating = 0
    recreational_within_1km_average_rating = 0
    education_within_1km_average_rating = 0
    healthcare_within_2km_average_rating = 0
    recreational_within_2km_average_rating = 0
    education_within_2km_average_rating = 0
    
    # Count healthcare within 1km radius
    healthcare_within_1km_count, healthcare_within_1km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                                 'Healthcare', 1) 
    # Count recreational areas within 1km radius
    recreational_within_1km_count, recreational_within_1km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                                     'Recreational', 1)
    # Count educational instituitions within 1km radius
    education_within_1km_count, education_within_1km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                               'Education', 1)
    # Count healthcare within 2km radius
    healthcare_within_2km_count, healthcare_within_2km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                                 'Healthcare', 2) 
    # Count recreational areas within 1km radius
    recreational_within_2km_count, recreational_within_2km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                                     'Recreational', 2)
    # Count educational instituitions within 1km radius
    education_within_2km_count, education_within_2km_average_rating = count_pois_within_radius(hdb_lng, hdb_lat, data_poi_filtered,
                                                                                               'Education', 2)
    
    # Add the recreational_count and school_count columns to the original hdb dataset
    street_avg_coords.loc[index, 'healthcare_within_1km_count'] = healthcare_within_1km_count
    street_avg_coords.loc[index, 'healthcare_within_1km_average_rating'] = healthcare_within_1km_average_rating
    
    street_avg_coords.loc[index, 'recreational_within_1km_count'] = recreational_within_1km_count
    street_avg_coords.loc[index, 'recreational_within_1km_average_rating'] = recreational_within_1km_average_rating
    
    street_avg_coords.loc[index, 'education_within_1km_count'] = education_within_1km_count
    street_avg_coords.loc[index, 'education_within_1km_average_rating'] = education_within_1km_average_rating
    
    street_avg_coords.loc[index, 'healthcare_within_2km_count'] = healthcare_within_2km_count
    street_avg_coords.loc[index, 'healthcare_within_2km_average_rating'] = healthcare_within_2km_average_rating
    
    street_avg_coords.loc[index, 'recreational_within_2km_count'] = recreational_within_2km_count
    street_avg_coords.loc[index, 'recreational_within_2km_average_rating'] = recreational_within_2km_average_rating
    
    street_avg_coords.loc[index, 'education_within_2km_count'] = education_within_2km_count
    street_avg_coords.loc[index, 'education_within_2km_average_rating'] = education_within_2km_average_rating

street_avg_coords

street_name    avg_long   avg_lat  healthcare_within_1km_count  \
0      ADMIRALTY DR  103.815851  1.450018                          1.0   
1    ADMIRALTY LINK  103.816254  1.455540                          1.0   
2        AH HOOD RD  103.844667  1.327701                         48.0   
3    ALJUNIED AVE 2  103.886335  1.319203                          9.0   
4     ALJUNIED CRES  103.884427  1.320423                          7.0   
..              ...         ...       ...                          ...   
561   YUNG KUANG RD  103.721629  1.331872                          1.0   
562     YUNG LOH RD  103.722128  1.328313                          1.0   
563    YUNG PING RD  103.722079  1.329695                          1.0   
564   YUNG SHENG RD  103.721681  1.333492                          1.0   
565         ZION RD  103.832608  1.290660                          8.0   

     healthcare_within_1km_average_rating  recreational_within_1km_count  \
0                                3.800000                            3.0   
1                                3.800000                            3.0   
2                                3.147917                           12.0   
3                                1.333333                            6.0   
4                                1.714286                            2.0   
..                                    ...                            ...   
561                              3.300000                            2.0   
562                              3.300000                            2.0   
563                              3.300000                            2.0   
564                              3.300000                            2.0   
565                              1.850000                            7.0   

     recreational_within_1km_average_rating  education_within_1km_count  \
0                                  3.733333                         4.0   
1                                  3.733333                         2.0   
2                                  4.016667                         4.0   
3                                  4.016667                         8.0   
4                                  4.300000                         6.0   
..                                      ...                         ...   
561                                4.250000                         2.0   
562                                4.250000                         2.0   
563                                4.250000                         2.0   
564                                4.250000                         3.0   
565                                4.057143                         3.0   

     education_within_1km_average_rating  healthcare_within_2km_count  \
0                               4.275000                          2.0   
1                               4.100000                          1.0   
2                               4.425000                         87.0   
3                               4.187500                         23.0   
4                               4.100000                         22.0   
..                                   ...                          ...   
561                             4.300000                         11.0   
562                             4.300000                          3.0   
563                             4.300000                          7.0   
564                             4.033333                         14.0   
565                             2.700000                        183.0   

     healthcare_within_2km_average_rating  recreational_within_2km_count  \
0                                3.900000                            6.0   
1                                3.800000                            4.0   
2                                2.542529                           16.0   
3                                1.708696                           15.0   
4                                1.722727       

In [37]:
# Merge counts to dataset
data_hdb_last15 = pd.merge(street_avg_coords, data_hdb_last15, on='street_name', how='left')

# Rearrange columns
cols = data_hdb_last15.columns.tolist()
data_hdb_last15 = data_hdb_last15[['town', 'flat_type', 'flat_model', 'floor_area_sqm', 'resale_price', 'month', 'lease_commence_date',
 'storey_range', 'block', 'remaining_lease', 'address', 'full_address', 'lat', 'long', 'nearest_mrt', 'nearest_distance_to_mrt',
 'flat_category', 'street_name', 'avg_long', 'avg_lat', 'healthcare_within_1km_count', 'healthcare_within_1km_average_rating', 
                                   'healthcare_within_2km_count','healthcare_within_2km_average_rating',
                                   'recreational_within_1km_count', 'recreational_within_1km_average_rating',
                                   'recreational_within_2km_count', 'recreational_within_2km_average_rating',
                                   'education_within_1km_count', 'education_within_1km_average_rating',
                                   'education_within_2km_count', 'education_within_2km_average_rating',
                                   'postal_code','Region']]
data_hdb_last15

town flat_type         flat_model  floor_area_sqm  \
0         SEMBAWANG    5 ROOM  Premium Apartment           111.0   
1         SEMBAWANG    5 ROOM  Premium Apartment           110.0   
2         SEMBAWANG    5 ROOM  Premium Apartment           110.0   
3         SEMBAWANG    4 ROOM  Premium Apartment            95.0   
4         SEMBAWANG    4 ROOM  Premium Apartment            95.0   
...             ...       ...                ...             ...   
325965  BUKIT MERAH    5 ROOM           Improved           114.0   
325966  BUKIT MERAH    5 ROOM           Improved           117.0   
325967  BUKIT MERAH    5 ROOM           Improved           117.0   
325968  BUKIT MERAH    5 ROOM           Improved           114.0   
325969  BUKIT MERAH    5 ROOM           Improved           117.0   

        resale_price      month lease_commence_date storey_range block  \
0           362000.0 2009-07-01          2001-01-01     04 TO 06  357A   
1           370000.0 2009-08-01          2001-01-01     07 TO 09  357A   
2           403000.0 2010-01-01          2001-01-01     16 TO 18  357A   
3           350000.0 2010-07-01          2001-01-01     01 TO 03  357A   
4           399000.0 2010-07-01          2001-01-01     10 TO 12  357A   
...              ...        ...                 ...          ...   ...   
325965      921000.0 2022-10-01          1974-01-01     13 TO 15    87   
325966      930000.0 2022-10-01          1974-01-01     04 TO 06    87   
325967      978000.0 2022-12-01          1974-01-01     13 TO 15    87   
325968      950000.0 2022-12-01          1974-01-01     22 TO 24    87   
325969      888000.0 2023-01-01          1974-01-01     04 TO 06    87   

           remaining_lease  ... recreational_within_1km_count  \
0                       91  ...                           3.0   
1                       91  ...                           3.0   
2                       90  ...                           3.0   
3                       90  ...                           3.0   
4                       90  ...                           3.0   
...                    ...  ...                           ...   
325965  50 years 09 months  ...                           7.0   
325966  50 years 09 months  ...                           7.0   
325967  50 years 08 months  ...                           7.0   
325968  50 years 07 months  ...                           7.0   
325969  50 years 06 months  ...                           7.0   

       recreational_within_1km_average_rating  recreational_within_2km_count  \
0                                    3.733333                            6.0   
1                                    3.733333                            6.0   
2                                    3.733333                            6.0   
3                                    3.733333                            6.0   
4                                    3.733333                            6.0   
...                                       ...                            ...   
325965                               4.057143                           78.0   
325966                               4.057143                           78.0   
325967                               4.057143                           78.0   
325968                               4.057143                           78.0   
325969                               4.057143                           78.0   

        recreational_within_2km_average_rating education_within_1km_count  \
0                                     3.916667                        4.0   
1                                     3.916667                        4.0   
2                                     3.916667                        4.0   
3                                     3.916667                        4.0   
4                                     3.916667                        4.0   
...                                        ...                        ...   
325965                      

In [28]:
# Check Merged Dataset
#data_hdb_last15.sample(10)
data_hdb_last15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 325970 entries, 0 to 325969
Data columns (total 28 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   town                           325970 non-null  object        
 1   flat_type                      325970 non-null  object        
 2   flat_model                     325970 non-null  object        
 3   floor_area_sqm                 325970 non-null  float64       
 4   resale_price                   325970 non-null  float64       
 5   month                          325970 non-null  datetime64[ns]
 6   lease_commence_date            325970 non-null  datetime64[ns]
 7   storey_range                   325970 non-null  object        
 8   block                          325970 non-null  object        
 9   remaining_lease                325970 non-null  object        
 10  address                        325970 non-null  object        
 11  

### Saving Dataset

In [38]:
# Save Merged Dataset
data_hdb_last15.to_csv('../dataset/hdb_last15_merged', index = False)